contrib/babelfishpg_tsql/src/procedures.c

```diff
@@ -1588,6 +1590,229 @@ create_xp_instance_regread_in_master_dbo_internal(PG_FUNCTION_ARGS)
+	const char *saved_dialect = GetConfigOption("babelfishpg_tsql.sql_dialect", true, true);
+	Oid			current_user_id = GetUserId();
+	const char *saved_buffer = pstrdup(GetConfigOption("search_path", true, true));
+	const char *new_buffer = "public, %s, \"$user\", sys, pg_catalog";
+	
+	PG_TRY();
+	{
+		set_config_option("babelfishpg_tsql.sql_dialect", "postgres",
+						GUC_CONTEXT_CONFIG,
+						PGC_S_SESSION, GUC_ACTION_SAVE, true, 0, false);
+		
+		postgresStmt = PG_ARGISNULL(0) ? NULL : TextDatumGetCString(PG_GETARG_TEXT_PP(0));
+
+		/* Remove trailing whitespaces */
+		len = strlen(postgresStmt);
+		while (isspace(postgresStmt[len - 1]))
+			postgresStmt[--len] = 0;
+
+		len = strlen(postgresStmt);
+
+		/* check if input statement is empty after removing trailing spaces */
+		if (len == 0)
+			ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+							errmsg("Statement cannot be NULL.")));
+
+		parsetree_list = raw_parser(postgresStmt, RAW_PARSE_DEFAULT);
+		stmt = parsetree_nth_stmt(parsetree_list, 0);
+		rewrite_object_refs(stmt);
+
+		if (list_length(parsetree_list) != 1)
+			ereport(ERROR,
+					(errcode(ERRCODE_SYNTAX_ERROR),
+					errmsg("Expected 1 statement but got %d statements after parsing.", list_length(parsetree_list))));
+
+        stmt = ((RawStmt *) linitial(parsetree_list))->stmt;
+			
+		/* need to make a wrapper PlannedStmt */
+		wrapper = makeNode(PlannedStmt);
+		wrapper->commandType = CMD_UTILITY;
+		wrapper->canSetTag = false;
+		wrapper->utilityStmt = stmt;
+		wrapper->stmt_location = 0;
+		wrapper->stmt_len = len;
+
+		parsetree = wrapper->utilityStmt;
+
+		switch (nodeTag(parsetree))
+		{
+			case T_CreateExtensionStmt:
+			{
+				CreateExtensionStmt *crstmt = (CreateExtensionStmt *) parsetree;
+				DefElem    *d_schema = NULL;
+				ListCell   *lc;
+				char	   *schemaName = NULL;
+
+				if (!superuser_arg(GetSessionUserId()) || !role_is_sa(GetSessionUserId()))
+				{
+					ereport(ERROR,
+						(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+	 					errmsg("permission denied to create extension")));
+				}
+
+				SetCurrentRoleId(GetSessionUserId(), false);
+				SetConfigOption("search_path",
+								new_buffer,
+								PGC_SUSET,
+								PGC_S_DATABASE_USER);
+
+				foreach(lc, crstmt->options)
+				{            
+					DefElem    *defel = (DefElem *) lfirst(lc);                                                                                                                                                                               
+					if (strcmp(defel->defname, "schema") == 0) 
+					{
+						d_schema = defel;
+						schemaName = defGetString(d_schema);
+					}
+					if (strcmp(defel->defname, "cascade") == 0)
+					{
+						ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+							errmsg("'cascade' is not yet supported in Babelfish.")));
+					}
+				}
+
+				if(schemaName != NULL && (strcmp(schemaName, "sys") && strcmp(schemaName, "public")))
```

Comment:
I think decision was to allow user to create extension in any schema of their choice. We will just document the recommendation. If there is confusion, please verify with Rob.